tests/unit/s2n_security_policies_test.c

```diff
@@ -54,6 +59,39 @@ static S2N_RESULT s2n_test_security_policies_compatible(const struct s2n_securit
     return S2N_RESULT_OK;
 }
 
+static S2N_RESULT s2n_test_default_backwards_compatible(const char *default_version,
+        const struct s2n_security_policy **versioned_policies, size_t versioned_policies_count,
+        const struct s2n_supported_cert *supported_certs, size_t supported_certs_count)
+{
+    RESULT_ENSURE_REF(default_version);
+    RESULT_ENSURE_REF(versioned_policies);
+    RESULT_ENSURE_REF(supported_certs);
+
+    /* The list of versioned policies MUST be kept up to date so that
+     * we continue testing against all past defaults.
+     */
+    const struct s2n_security_policy *current = NULL;
+    RESULT_GUARD_POSIX(s2n_find_security_policy_from_version(default_version, &current));
+    if (versioned_policies[versioned_policies_count - 1] != current) {
+        fprintf(stdout, "Missing latest version of '%s'\n", default_version);
+        FAIL_MSG("New default policy MUST be added to versioning test");
+    }
+
+    for (size_t policy_i = 0; policy_i < versioned_policies_count; policy_i++) {
+        for (size_t cert_i = 0; cert_i < supported_certs_count; cert_i++) {
+            if (policy_i < supported_certs[cert_i].start_index) {
+                continue;
+            }
+            RESULT_GUARD(s2n_test_security_policies_compatible(
+                    versioned_policies[policy_i],
+                    default_version,
+                    supported_certs[cert_i].cert));
```

Comment:
This seems like it might be infeasibly restrictive. I agree that `policy[i]` should always be able to negotiate with `policy[i - 1]`, but I don't think we should commit to `policy[0]` being able to negotiate with `policy[n]`

Reply:
The problem is that customers aren't guaranteed to deploy every version of s2n-tls in order. It's not that we can NEVER violate this "compatible with all previous versions" promise, but we'll need to know and acknowledge that violating it is a big deal and could break customers who update really rarely.